Login Rework: email screen visual touches

hypest opened the pull request:
**Note: #6196 needs to be merged first.**

This PR is about the visual touches for the email input screen.

On the code side of things, the `TextInputEditLayout` got subclassed into `WPTextInputEditLayout` in order to fix 2 issues of the original: 1) Support baseline computation so icon on the left can align to it and 2) excess side padding of the error TextView. Also, the `TextInputEditText` used is translated horizontally to remove its excess side padding.

The "input row" (icon and email text input) is defined as a custom View for reusability. Ideally, it will be completely replaced by `TextInputEditLayout` when the Support Lib is updated to fully follow Material Design standard.

To test:
* With the app data cleaned, launch the app and hit the "Log in" button
* Visually assess the email screen

hypest committed:
Reintroduced reader prologue anim, tighter canvas

hypest committed:
Horizontal bottom buttons in landscape prologue

hypest committed:
Center promo text vertically as well

hypest committed:
Working reader.json

hypest committed:
Alternate light and darker prologue pages

hypest committed:
Use Material Title for promo text

hypest committed:
Use standard Material buttons in prologue

hypest committed:
24dp visible bottom, top gap as per login design

hypest committed:
8dp margin bewteen touch targets as per Material

hypest committed:
Add a shadow from bottom buttons to anims

hypest committed:
Space out the prologue screen elements

hypest committed:
Do 2 anim loops before auto-advancing

hypest committed:
Fix layout lint warnings on prologue screens

hypest committed:
Add missing "a" from the button text

hypest committed:
Adjust the side margins of promo text

hypest committed:
Icon outside of TextInputLayout and baselined

hypest committed:
Merge branch 'feature/new-login' into feature/login-prologue-updated-anims

hypest committed:
grey_lighten_10 for the icon

hypest committed:
Use Material Body1 for body text

hypest committed:
Set body background to grey_light

hypest committed:
Set TextInputLayout's colors

hypest committed:
Remove useless side padding of TextInputEditLayout

hypest committed:
Set input text style to Body1

hypest committed:
16dp margin between icon and input as per Material

hypest committed:
Leave Next button enabled for discoverability

Empty email input will produce an error (and error message) anyway.

hypest committed:
Visual touches and resusable input row layout file

hypest committed:
Use the correct id

folletto commented:
Can you add screenshots? Thanks! 😀

hypest committed:
Translucent status bar

hypest committed:
Just one animation loop before auto-advance

hypest committed:
Reset the animation on each screen when swiping

hypest committed:
Upgrade Gradle wrapper to v3.5

hypest committed:
Upgrade Gradle wrapper to v3.5

hypest committed:
Upgrade Bintray plugin to v0.5.0

This also the problem where Bintray was using a private API and build failed.

hypest committed:
Upgrade Bintray plugin to v0.5.0

This also the problem where Bintray was using a private API and build failed.

hypest committed:
Pick up gradle.properties from BuddyBuild if avail

hypest committed:
Pick up gradle.properties from BuddyBuild if avail

hypest committed:
Use em-dash instead of hyphen

hypest committed:
Stop auto-advancing if user swiped manually

hypest committed:
Merge branch 'develop' into feature/login-prologue-updated-anims

hypest committed:
Reinstate static Status Bar color

hypest commented:
Hey @folletto , I just enabled BuddyBuild on this PR. Hope it helps with the visual review, thanks!

hypest committed:
Merge branch 'feature/new-login' into feature/login-prologue-updated-anims

hypest committed:
Merge branch 'feature/new-login' into feature/login-visual-touches-email

folletto commented:
> I just enabled BuddyBuild on this PR. Hope it helps with the visual review. 

BB is awesome, thanks! However the screenshot isn't just to get a fast review (sometimes the screenshot is enough, sometimes the full build is better) but also for future use to see what change was done by a specific PR. :)

hypest committed:
Bottom buttons frame, form content panning

When the virtual keyboard opens the bottom buttons will correctly shadow the main form contents while the contents also try to better position in the scrollview. Especially useful when small screen.

hypest commented:
Fixed (d8a909c) an issue with the bottom buttons overlapping the main form elements when opening the virtual keyboard on a sufficiently small screen/device.

folletto commented:
Screenshots from this PR:

| Not selected | Selected |
| --- | --- |
|  ![screenshot_20170710-141018](https://user-images.githubusercontent.com/4389/28019758-f7f98ff2-6579-11e7-90ce-256994c766cf.png) | ![screenshot_20170710-141022](https://user-images.githubusercontent.com/4389/28019767-fefbf6fa-6579-11e7-8024-c903532f73e3.png) |

folletto commented:
I think the design is good. There are just some minor differences, but I think it looks good overall and I will be ok in pushing this.

Tiny things and confirmations:

1. If I recall correctly we can't do the full underline (including the icon), is that right?
2. The "Log in to your site by entering..." is slightly tighter in terms of line height. I reckon is standard on all screens, so I'm ok in having it consistent.
3. There's a full stop "." missing at the end of the first sentence.

hypest commented:
> If I recall correctly we can't do the full underline (including the icon), is that right?

The problem lies with the password field (not in this screen though; there's no password input here) that hides the icon. So, to cater for that case we put the icon on the outside for all input fields. Besides, Material Design has the icon on the outside anyway. See "Icon signifier" in https://material.io/guidelines/components/text-fields.html#text-fields-principles for reference.

> The "Log in to your site by entering..." is slightly tighter in terms of line height. I reckon is standard on all screens, so I'm ok in having it consistent.

Ah, that's fairly easy to configure so, I will set it to `2dp` (same as the promo texts).

> There's a full stop "." missing at the end of the first sentence.

Oops, will add it, good catch!

hypest committed:
Add some line spacing in the form text

folletto commented:
> So, to cater for that case we put the icon on the outside for all input fields. Besides, Material Design has the icon on the outside anyway.

Yep. I was just double checking the decision, thanks!

👍  on everything else.

hypest committed:
Add missing period

hypest committed:
More generic login form layout file

hypest committed:
Introduce a base fragment for main login screens

hypest commented:
Hey @folletto , I made a small UI change to decrease the margin above the screen text to perform better with small screens (3.7") when virtual keyboard opens. You might want to have a look.

Here's a screenshot of the new margin:
![shorter_text_top_margin](https://user-images.githubusercontent.com/1032913/28027715-cf5164c0-65a2-11e7-919c-98a50f015e1b.jpg)

hypest committed:
Remove OverridingMethodsMustInvokeSuper

As it breaks the vanillaRelease build.

folletto commented:
Nice!

Question: would be possible to pre-focus on the field, and have the keyboard open already?
I felt this already doing the tests, and by moving the items up it seems even more evident now.

hypest commented:
> Question: would be possible to pre-focus on the field, and have the keyboard open already?

Yeap! Makes sense. I have that in some other screen but forgot to add it here.

hypest committed:
Focus email input box on screen open

hypest commented:
@folletto , implemented auto-focus of the input field upon entering the screen. Feel free to have a look. Thanks!

hypest committed:
White action text color for older APIs as well

hypest committed:
Hide keyboard on prologue screen

Needed when returning from the email screen with the keyboard visible.

hypest committed:
Minor code cleanup

hypest committed:
Remove promo anims auto-advancing

hypest committed:
Reset anim on promo anim page hide

folletto commented:
> implemented auto-focus of the input field upon entering the screen. Feel free to have a look. Thanks!

Tested — works really well 👍

hypest committed:
Rename input row layout to a more generic name

hypest committed:
Move method closer to sibling ones

hypest committed:
Merge branch 'feature/new-login' into feature/login-prologue-updated-anims

hypest committed:
Merge branch 'feature/new-login' into feature/login-visual-touches-email

hypest committed:
Custom View for the login input row

hypest committed:
Merge branch 'feature/new-login' into feature/login-prologue-updated-anims

hypest committed:
Merge branch 'feature/new-login' into feature/login-visual-touches-email

hypest committed:
Dedicated message for deeplink login

hypest committed:
Update to newer Stats, Notifications anims

hypest commented:
FYI @theck13 , I've introduced a custom View (`WPLoginInputRow`) to engulf the visual style and functionality we need for the input boxes. Unfortunately, the `TextInputLayout` provided by the Support lib does not adhere to the Material Design standard we want so, this custom View will be used in place until Google updates the Support Lib.

hypest committed:
Cleaner setting of editor "enter" button listener

hypest committed:
Capture imeAction when done/next as button commit

hypest committed:
Allow customizing the EditText's imeOptions

hypest committed:
Set virtual keyboards action to "Next" for email

hypest committed:
Minor code cleanup

hypest committed:
Set only if attribute is present in the layout xml

hypest committed:
Make sure to save rules in LayoutParams struct

hypest committed:
Change layout rules directly in array for API >23

hypest committed:
Code cleanup

hypest committed:
Minor codestyle fix

hypest committed:
Add missing shadow in landscape

hypest committed:
Merge branch 'feature/new-login' into feature/login-prologue-updated-anims

hypest committed:
Merge branch 'feature/login-prologue-updated-anims' into feature/login-visual-touches-email

theck13 commented on WordPress/src/main/res/values/attrs.xml:
```diff
@@ -107,4 +107,17 @@
+        <attr name="android:imeOptions" />
+        <attr name="passwordToggleEnabled" />
+        <attr name="passwordToggleTint" />
+    </declare-styleable>
```
Should `passwordToggleEnabled` have `format="boolean"` and `passwordToggleTint ` have `format="reference|color"`?

hypest replied:
Not really, the formats get inherited from the design support lib's definitions.

theck13 replied:
Oops, I didn't realize `passwordToggleEnabled` and `passwordToggleTint` were inherited attributes.

theck13 commented on WordPress/src/main/res/values/attrs.xml:
```diff
@@ -107,4 +107,17 @@
+    <declare-styleable name="wpLoginInputRow">
+        <attr name="wpIconDrawable" format="reference"/>
+        <attr name="wpIconContentDescription" format="string"/>
+        <attr name="android:inputType" />
```
Should the `wp` be removed from `wpIconDrawable` and `wpIconContentDescription` to be consistent with the other `wpLoginInputRow` attributes?

hypest replied:
Hmm, the other attributes are actually inherited from Android's set of attributes and are used as normal. The `wp` prefixed ones are basically the "real" custom ones and that fact becomes clearer when using them. I'd suggest we keep the different naming. WDYT?

theck13 replied:
Ahh, I didn't realize `passwordToggleEnabled` and `passwordToggleTint` were inherited from `TextInputLayout`.

theck13 commented on WordPress/src/main/java/org/wordpress/android/widgets/WPLoginInputRow.java:
```diff
@@ -0,0 +1,236 @@
+        mEditText = (EditText) findViewById(R.id.input);
+
+        if (attrs != null) {
+            TypedArray a = context.getTheme().obtainStyledAttributes(attrs, R.styleable.wpLoginInputRow, 0, 0);
```
What about using an `else` condition or a ternary expression for each attribute so that a default value is assigned if no custom value exists?  For methods like `setImageResource` and `getBoolean` that require a default value, no conditional statements or ternary expressions are needed.

hypest replied:
I intentionally left out the `else`s so the default attributes of the components are applied anyway. I'd suggest we introduce `else`s on a case-by-case manner when needed.

By the way, in a different PR (https://github.com/wordpress-mobile/WordPress-Android/pull/6341#issuecomment-316314210) I did introduce an `else` case for the icon since we specifically want some behaviour to happen on the `else`.

theck13 replied:
That's fine.

theck13 commented on WordPress/src/main/java/org/wordpress/android/widgets/WPLoginInputRow.java:
```diff
@@ -0,0 +1,236 @@
+        rules = editTextLayoutParams.getRules();
+        for (int i = 0; i < rules.length; i++) {
+            if (rules[i] == mIcon.getId()) {
+                if (Build.VERSION.SDK_INT > 23) {
```
The constant `Build.VERSION_CODES.M` should be used instead of hardcoding `23`.

hypest replied:
Good catch! Fixed with 4ca9473.

theck13 commented on WordPress/src/main/java/org/wordpress/android/widgets/WPLoginInputRow.java:
```diff
@@ -0,0 +1,236 @@
+        int[] rules = iconLayoutParams.getRules();
+        for (int i = 0; i < rules.length; i++) {
+            if (rules[i] == mTextInputLayout.getId()) {
+                if (Build.VERSION.SDK_INT > 23) {
```
The constant `Build.VERSION_CODES.M` should be used instead of hardcoding `23`.

hypest replied:
Good catch! Fixed with 4ca9473.

theck13 commented on WordPress/src/main/java/org/wordpress/android/widgets/WPLoginInputRow.java:
```diff
@@ -0,0 +1,236 @@
+            mIds = ids;
+        }
+
+        private SavedState(Parcel in) {
```
This `SavedState` constructor can be made package-private to avoid accessing a private member in the `createFromParcel` method.

hypest replied:
Thanks! Aligned it with the other constructor already defined (9ecfabb).

theck13 commented on WordPress/src/main/java/org/wordpress/android/widgets/WPTextInputLayout.java:
```diff
@@ -0,0 +1,46 @@
+
+    @Override
+    public int getBaseline()
+    {
```
This curly brace should be on the previous line for style consistency.

hypest replied:
Argh, of course! Fixed with 764fe9b.

theck13 commented on WordPress/src/main/res/values/dimens.xml:
```diff
@@ -70,7 +70,9 @@
     <dimen name="margin_small">4dp</dimen>
+    <dimen name="margin_small_medium">6dp</dimen>
     <dimen name="margin_medium">8dp</dimen>
+    <dimen name="margin_extra_medium">10dp</dimen>
```
Should `margin_extra_medium` be refactored to `margin_medium_large` to stay consistent with the `margin_small_medium` naming convention?

hypest replied:
I was mirroring the naming of `margin_extra_large` vs `margin_large` but your suggestion makes more sense! Fixed with d99d0a3.

hypest commented on WordPress/src/main/java/org/wordpress/android/ui/accounts/login/LoginEmailFragment.java:
```diff
@@ -31,81 +20,79 @@
-        mEmailEditText.addTextChangedListener(this);
+        if (mLoginListener.getLoginMode() == LoginMode.JETPACK_STATS) {
+            label.setText(R.string.stats_sign_in_jetpack_different_com_account);
+        } else if (mLoginListener.getLoginMode() == LoginMode.JETPACK_STATS) {
```
Argh, a merge hiccup. Will fix!

hypest commented on WordPress/src/main/java/org/wordpress/android/ui/accounts/login/LoginEmailFragment.java:
```diff
@@ -31,81 +20,79 @@
+        } else if (mLoginListener.getLoginMode() == LoginMode.JETPACK_STATS) {
+
+        } else {
+            label.setText(R.string.enter_email_wordpress_com);
```
Yeap, the `switch` covers it. Will fix this merge hiccup as well.

hypest committed:
Code cleanup

hypest commented on WordPress/src/main/java/org/wordpress/android/ui/accounts/login/LoginBaseFormFragment.java:
```diff
@@ -0,0 +1,195 @@
+                        new DialogInterface.OnCancelListener() {
+                            @Override
+                            public void onCancel(DialogInterface dialogInterface) {
+                                if (mInProgress) {
```
I'm not seeing a warning in AS. Is that some `lint` warning perhaps?

I'd be inclined to change it to use the getter btw.

theck13 replied:
The `if (mInProgress)` statement was changed to `if (isInProgress())` in 33673d5b6361df192fdb8b739fffd1add4ae6d71 so the lint warning is no longer visible.  I'm fine with leaving `isInProgress()` if you want.

hypest committed:
Rename to startProgress to complement endProgress

hypest commented on WordPress/src/main/java/org/wordpress/android/ui/accounts/login/LoginEmailFragment.java:
```diff
@@ -1,24 +1,13 @@
 import android.text.TextUtils;
 import android.text.TextWatcher;
 import android.util.Patterns;
 import android.view.KeyEvent;
```
Done with 4b5f0b5.

hypest commented on WordPress/src/main/java/org/wordpress/android/ui/accounts/login/LoginBaseFormFragment.java:
```diff
@@ -0,0 +1,195 @@
+        mInProgress = true;
+    }
+
+    protected void endProgress() {
```
Done with 33673d5.

hypest committed:
Use the contant instead of the literal

hypest committed:
Use the contant here as well

hypest committed:
Mark as package-private like the other constructor

hypest committed:
Codestyle fix

hypest committed:
Mirror the naming of `margin_small_medium`

hypest commented:
I addressed your comments @theck13 , ready for another pass! Thanks!

hypest committed:
Better hierarchical naming for login button styles

hypest committed:
Remove unneeded color setting

hypest committed:
Merge branch 'feature/login-prologue-updated-anims' into feature/login-visual-touches-email

hypest committed:
Use the hierarchical naming for Button.Secondary

theck13 reviewed (approved):
The changes look good!  I'll wait for https://github.com/wordpress-mobile/WordPress-Android/pull/6196 before merging this one.

hypest committed:
Don't clip bottom button shadows in form screens

theck13 reviewed (approved):
:shipit:
